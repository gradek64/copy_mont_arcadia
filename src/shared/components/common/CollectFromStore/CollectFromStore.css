/** @define CollectFromStore */

.CollectFromStore {
  display: flex;
  flex-direction: column;
  position: absolute;
  left: $page-padding;
  right: $page-padding;
  top: $page-padding;
  bottom: $page-padding;

  &-row {
    align-items: stretch;
    display: flex;
    flex: 1;
    min-height: 0;
  }

  &-columnLeft {
    display: flex;
    flex: .55;
    flex-direction: column;
    margin-right: 10px;
  }

  &-columnRight {
    flex: .45;
    position: relative;
  }

  &-espotContainer {
    background-color: $store-locator-espot-background-color;
    margin-top: $page-padding;
    width: 100%;
  }

  &-listContainer {
    display: flex;
    flex: 1;
    flex-direction: column;
    margin-top: $page-padding;
    min-height: 0;
  }

  &-overflowContainer {
    border: 1px solid $collect-from-store-list-boder-color;
    flex: 1;
    overflow: auto;
  }
  
  &-filters {
    background-color: $collect-from-store-filter-bg-color;
  }
}

/* postcss-bem-linter: ignore */
.CollectFromStore .GoogleMap {
  position: absolute;
  left: 0;
  right: 0;
  top: 0;
  bottom: 0;
}
