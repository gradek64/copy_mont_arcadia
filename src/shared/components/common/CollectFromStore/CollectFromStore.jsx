import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { isEmpty } from 'ramda'

import UserLocatorInput from '../UserLocatorInput/UserLocatorInput'
import GoogleMap from '../StoreLocator/GoogleMap'
import StoreList from '../StoreLocator/StoreList'

import { getStoresForCheckoutModal } from '../../../actions/components/StoreLocatorActions'
import {
  selectDeliveryStore,
  setStoreUpdating,
} from '../../../actions/common/checkoutActions'
import { closeModal } from '../../../actions/common/modalActions'

import { getAppliedFilters } from '../../../lib/store-locator-utilities'
import storeLocatorConsts from '../../../constants/storeLocator'
import Form from '../FormComponents/Form/Form'
import { getMapCentrePointAndZoom } from '../../../selectors/storeLocatorSelectors'
import { getUserInputGeoLocation } from '../../../selectors/userLocatorSelectors'

@connect(
  (state) => ({
    stores: state.storeLocator.stores,
    storeQuery: state.storeLocator.query,
    filters: state.storeLocator.filters,
    isStoresLoading: state.storeLocator.loading,
    isMobile: state.viewport.media === 'mobile',
    selectedPlaceDetails: state.userLocator.selectedPlaceDetails,
    mapCentrePointAndZoom: getMapCentrePointAndZoom(state),
    getUserInputGeoLocation: getUserInputGeoLocation(state),
  }),
  {
    getStoresForCheckoutModal,
    selectDeliveryStore,
    setStoreUpdating,
    closeModal,
  }
)
class CollectFromStore extends Component {
  static propTypes = {
    isMobile: PropTypes.bool.isRequired,
    stores: PropTypes.array.isRequired,
    isStoresLoading: PropTypes.bool.isRequired,
    filters: PropTypes.object.isRequired,
    storeQuery: PropTypes.object,
    selectedPlaceDetails: PropTypes.object,
    getStoresForCheckoutModal: PropTypes.func,
    selectDeliveryStore: PropTypes.func,
    setStoreUpdating: PropTypes.func,
    mapCentrePointAndZoom: PropTypes.object,
    getUserInputGeoLocation: PropTypes.bool,
  }

  static contextTypes = {
    l: PropTypes.func,
  }
  state = {
    dimensions: null,
    noStoresFound: false,
  }

  UNSAFE_componentWillMount() {
    const {
      storeQuery,
      filters,
      getStoresForCheckoutModal,
      selectedPlaceDetails,
      getUserInputGeoLocation,
    } = this.props

    if (
      storeQuery !== undefined &&
      getAppliedFilters(filters).join(',') !== storeQuery.types &&
      (!isEmpty(selectedPlaceDetails) || getUserInputGeoLocation)
    ) {
      getStoresForCheckoutModal()
    }
  }

  UNSAFE_componentWillReceiveProps(nextProps) {
    const { isMobile, closeModal } = this.props
    if (nextProps.isMobile && isMobile !== nextProps.isMobile) closeModal()
  }

  componentDidUpdate(prevProps) {
    if (this.props.stores !== prevProps.stores) {
      this.setState({ noStoresFound: this.props.stores.length === 0 }) // eslint-disable-line
    }
  }

  componentWillUnmount() {
    const { setStoreUpdating } = this.props
    setStoreUpdating(false)
  }

  onSelectDeliveryStore = (store) => {
    const { selectDeliveryStore, closeModal } = this.props
    selectDeliveryStore(store)

    closeModal()
  }

  searchStores = (e) => {
    const { getStoresForCheckoutModal } = this.props
    if (e) e.preventDefault()

    getStoresForCheckoutModal()
  }

  render() {
    const {
      stores,
      isMobile,
      isStoresLoading,
      mapCentrePointAndZoom: { lat, long, zoom, markers, iconDomain },
    } = this.props
    const { dimensions, noStoresFound } = this.state
    const { l } = this.context
    return (
      <div className="CollectFromStore">
        <h3 className="CollectFromStore-title">{l`Where do you want to collect from?`}</h3>
        <div className="CollectFromStore-row">
          <div className="CollectFromStore-columnLeft">
            <Form
              className="CollectFromStore-search"
              onSubmit={this.searchStores}
            >
              <UserLocatorInput
                selectedCountry={storeLocatorConsts.defaultCountry}
                submitHandler={this.searchStores}
              />
            </Form>
            {noStoresFound && !isStoresLoading ? (
              <p>
                {l`Sorry, we couldn't find any locations matching your search. Please enter another location and try again.`}
              </p>
            ) : null}
            {!!stores.length && !isStoresLoading ? (
              <StoreList
                name="CollectFromStore"
                storeType="collectFromStore"
                hasFilters
                fetchStores={this.searchStores}
                selectDeliveryStore={this.onSelectDeliveryStore}
              />
            ) : null}
          </div>
          <div
            className="CollectFromStore-columnRight"
            ref={(columnRight) => {
              if (columnRight && !dimensions) {
                this.setState({
                  dimensions: {
                    width: columnRight.offsetWidth,
                    height: columnRight.offsetHeight,
                  },
                })
              }
            }}
          >
            {process.browser &&
              !isMobile &&
              !isStoresLoading &&
              !!dimensions && (
                <GoogleMap
                  currentLat={lat}
                  currentLng={long}
                  markers={markers}
                  dimensions={dimensions}
                  iconDomain={iconDomain}
                  zoom={zoom}
                />
              )}
          </div>
        </div>
        {!stores.length || isStoresLoading ? (
          <div className="CollectFromStore-espotContainer" />
        ) : null}
      </div>
    )
  }
}

export default CollectFromStore
