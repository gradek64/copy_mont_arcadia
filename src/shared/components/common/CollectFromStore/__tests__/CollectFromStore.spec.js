import testComponentHelper from 'test/unit/helpers/test-component'
import CollectFromStore from '../CollectFromStore'

const mockedStoreList = [
  {
    brandId: 12556,
    distance: '0.16',
    latitude: 51.509599,
    longitude: -0.123069,
    name: 'Strand',
    storeId: 'TS0032',
  },
  {
    brandId: 45655,
    distance: '1.16',
    latitude: 51.439599,
    longitude: -0.123069,
    name: 'Strand 2',
    storeId: 'TS0033',
  },
  {
    brandId: 56784,
    distance: '1.56',
    latitude: 51.639599,
    longitude: -0.123069,
    name: 'Strand 3',
    storeId: 'TS0034',
  },
]
const mockStorefilters = {
  brand: {
    selected: true,
    applied: true,
  },
  other: {
    selected: true,
    applied: false,
  },
  parcel: {
    selected: false,
    applied: true,
  },
}

describe('<CollectFromStore/>', () => {
  const initialProps = {
    isMobile: false,
    stores: mockedStoreList,
    isStoresLoading: false,
    filters: mockStorefilters,
    getStoresForCheckoutModal: jest.fn(),
    selectDeliveryStore: jest.fn(),
    selectDeliveryStoreForQubit: jest.fn(),
    closeModal: jest.fn(),
    setStoreUpdating: jest.fn(),
    mapCentrePointAndZoom: {
      lat: 2,
      long: 1,
      zoom: 15,
      markers: [[1, 2], [3, 4]],
      iconDomain: 'static.brand.com',
    },
  }
  const renderComponent = testComponentHelper(
    CollectFromStore.WrappedComponent,
    { disableLifecycleMethods: false }
  )

  describe('@renders', () => {
    it('in default state', () => {
      expect(renderComponent(initialProps).getTree()).toMatchSnapshot()
    })
    it('when stores is empty', () => {
      expect(
        renderComponent({
          ...initialProps,
          stores: [],
        }).getTree()
      ).toMatchSnapshot()
    })
    it('should render no results message', () => {
      const { wrapper, getTree } = renderComponent(initialProps)

      wrapper.setProps({
        stores: [],
      })

      expect(getTree()).toMatchSnapshot()
    })
    it('when stores exist and getStoresLoading is true', () => {
      expect(
        renderComponent({
          ...initialProps,
          isStoresLoading: true,
        }).getTree()
      ).toMatchSnapshot()
    })
    it('when stores exist and getStoresLoading is false', () => {
      expect(renderComponent(initialProps).getTree()).toMatchSnapshot()
    })
  })

  describe('@lifecycle', () => {
    beforeEach(() => {
      jest.resetAllMocks()
    })
    describe('UNSAFE_componentWillMount', () => {
      it('should not call getStoresForCheckoutModal if storeQuery is not defined', () => {
        const { instance } = renderComponent({
          ...initialProps,
          selectedPlaceDetails: {
            storeId: 2344,
          },
        })
        expect(instance.props.getStoresForCheckoutModal).not.toHaveBeenCalled()
      })
      it('should not call getStoresForCheckoutModal if filters have already been applied ie the storeQuery type matches the filter', () => {
        const { instance } = renderComponent({
          ...initialProps,
          storeQuery: {
            types: 'brand,parcel',
          },
          selectedPlaceDetails: {
            storeId: 2344,
          },
        })
        expect(instance.props.getStoresForCheckoutModal).not.toHaveBeenCalled()
      })
      it('should not call getStoresForCheckoutModal if selectedPlaceDetails is empty', () => {
        const { instance } = renderComponent({
          ...initialProps,
          storeQuery: {
            types: 'brand',
          },
          selectedPlaceDetails: {},
        })
        expect(instance.props.getStoresForCheckoutModal).not.toHaveBeenCalled()
      })
      it('should call getStoresForCheckoutModal if storeQuery is defined and the filters have not yet been applied and selectedPlaceDetails is not empty', () => {
        const { instance } = renderComponent({
          ...initialProps,
          storeQuery: {
            types: 'brand',
          },
          selectedPlaceDetails: {
            storeId: 2344,
          },
        })
        jest.resetAllMocks()
        expect(instance.props.getStoresForCheckoutModal).not.toHaveBeenCalled()
        instance.UNSAFE_componentWillMount()
        expect(instance.props.getStoresForCheckoutModal).toHaveBeenCalledTimes(
          1
        )
      })
    })
    describe('UNSAFE_componentWillReceiveProps', () => {
      it('should call closeModal when changing media from desktop to mobile', () => {
        const { wrapper, instance } = renderComponent(initialProps)
        expect(instance.props.closeModal).not.toHaveBeenCalled()
        wrapper.setProps({
          isMobile: true,
        })
        expect(instance.props.closeModal).toHaveBeenCalledTimes(1)
      })
      it('should not call closeModal when the media does not change', () => {
        const { wrapper, instance } = renderComponent(initialProps)
        wrapper.setProps({
          isStoresLoading: true,
        })
        expect(instance.props.closeModal).not.toHaveBeenCalled()
      })
      it('should not call closeModal when the media changes to anything other than mobile', () => {
        const { wrapper, instance } = renderComponent({
          ...initialProps,
          isMobile: true,
        })
        wrapper.setProps({
          isMobile: false,
        })
        expect(instance.props.closeModal).not.toHaveBeenCalled()
      })
    })
    describe('componentDidUpdate', () => {
      it('should update `noStoresFound` state prop when stores are fetched', () => {
        const { wrapper } = renderComponent(initialProps)

        expect(wrapper.state('noStoresFound')).toBe(false)
        wrapper.setProps({
          stores: [],
        })

        expect(wrapper.state('noStoresFound')).toBe(true)
      })
    })
    describe('componentWillUnmount', () => {
      it('should call setStoreUpdating with false', () => {
        const { instance, wrapper } = renderComponent(initialProps)
        expect(instance.props.setStoreUpdating).not.toHaveBeenCalled()
        wrapper.unmount()
        expect(instance.props.setStoreUpdating).toHaveBeenCalledTimes(1)
        expect(instance.props.setStoreUpdating).toHaveBeenLastCalledWith(false)
      })
    })
  })

  describe('@events', () => {
    const event = {
      preventDefault: jest.fn(),
    }
    beforeEach(() => {
      jest.resetAllMocks()
    })
    describe('search form onSubmit', () => {
      it('should call getStoresForCheckoutModal() with "collectFromStore" and preventDefault', () => {
        const { wrapper, instance } = renderComponent({
          ...initialProps,
          analyticsReceiveStores: jest.fn(),
        })
        expect(instance.props.getStoresForCheckoutModal).not.toHaveBeenCalled()
        expect(event.preventDefault).not.toHaveBeenCalled()
        wrapper.find('Form').simulate('submit', event)
        expect(instance.props.getStoresForCheckoutModal).toHaveBeenCalledTimes(
          1
        )
        expect(event.preventDefault).toHaveBeenCalledTimes(1)
      })
    })
    describe('storelist', () => {
      it('on calling fetchStores should call getStoresForCheckoutModal() with "collectFromStore" and preventDefault', () => {
        const { wrapper, instance } = renderComponent({
          ...initialProps,
          analyticsReceiveStores: jest.fn(),
        })
        expect(instance.props.getStoresForCheckoutModal).not.toHaveBeenCalled()
        wrapper
          .find('Connect(StoreList)')
          .getElement()
          .props.fetchStores()
        expect(instance.props.getStoresForCheckoutModal).toHaveBeenCalledTimes(
          1
        )
        expect(event.preventDefault).not.toHaveBeenCalled()
      })
      it('on calling selectDeliveryStore should call selectDeliveryStore() and closeModal()', () => {
        const { wrapper, instance } = renderComponent(initialProps)
        const store = mockedStoreList[2]
        expect(instance.props.selectDeliveryStore).not.toHaveBeenCalled()
        expect(instance.props.closeModal).not.toHaveBeenCalled()
        wrapper
          .find('Connect(StoreList)')
          .getElement()
          .props.selectDeliveryStore(store)
        expect(instance.props.selectDeliveryStore).toHaveBeenCalledTimes(1)
        expect(instance.props.selectDeliveryStore).toHaveBeenLastCalledWith(
          store
        )
        expect(instance.props.closeModal).toHaveBeenCalledTimes(1)
      })
    })
  })
})
