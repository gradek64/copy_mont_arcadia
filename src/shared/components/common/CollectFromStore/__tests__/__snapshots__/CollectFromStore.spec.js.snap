// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CollectFromStore/> @renders in default state 1`] = `
<div
  className="CollectFromStore"
>
  <h3
    className="CollectFromStore-title"
  >
    Where do you want to collect from?
  </h3>
  <div
    className="CollectFromStore-row"
  >
    <div
      className="CollectFromStore-columnLeft"
    >
      <Form
        className="CollectFromStore-search"
        noValidate={true}
        onSubmit={[Function]}
      >
        <Connect(UserLocatorInput)
          selectedCountry="United Kingdom"
          submitHandler={[Function]}
        />
      </Form>
      <Connect(StoreList)
        fetchStores={[Function]}
        hasFilters={true}
        name="CollectFromStore"
        selectDeliveryStore={[Function]}
        storeType="collectFromStore"
      />
    </div>
    <div
      className="CollectFromStore-columnRight"
    />
  </div>
</div>
`;

exports[`<CollectFromStore/> @renders should render no results message 1`] = `
<div
  className="CollectFromStore"
>
  <h3
    className="CollectFromStore-title"
  >
    Where do you want to collect from?
  </h3>
  <div
    className="CollectFromStore-row"
  >
    <div
      className="CollectFromStore-columnLeft"
    >
      <Form
        className="CollectFromStore-search"
        noValidate={true}
        onSubmit={[Function]}
      >
        <Connect(UserLocatorInput)
          selectedCountry="United Kingdom"
          submitHandler={[Function]}
        />
      </Form>
      <p>
        Sorry, we couldn't find any locations matching your search. Please enter another location and try again.
      </p>
    </div>
    <div
      className="CollectFromStore-columnRight"
    />
  </div>
  <div
    className="CollectFromStore-espotContainer"
  />
</div>
`;

exports[`<CollectFromStore/> @renders when stores exist and getStoresLoading is false 1`] = `
<div
  className="CollectFromStore"
>
  <h3
    className="CollectFromStore-title"
  >
    Where do you want to collect from?
  </h3>
  <div
    className="CollectFromStore-row"
  >
    <div
      className="CollectFromStore-columnLeft"
    >
      <Form
        className="CollectFromStore-search"
        noValidate={true}
        onSubmit={[Function]}
      >
        <Connect(UserLocatorInput)
          selectedCountry="United Kingdom"
          submitHandler={[Function]}
        />
      </Form>
      <Connect(StoreList)
        fetchStores={[Function]}
        hasFilters={true}
        name="CollectFromStore"
        selectDeliveryStore={[Function]}
        storeType="collectFromStore"
      />
    </div>
    <div
      className="CollectFromStore-columnRight"
    />
  </div>
</div>
`;

exports[`<CollectFromStore/> @renders when stores exist and getStoresLoading is true 1`] = `
<div
  className="CollectFromStore"
>
  <h3
    className="CollectFromStore-title"
  >
    Where do you want to collect from?
  </h3>
  <div
    className="CollectFromStore-row"
  >
    <div
      className="CollectFromStore-columnLeft"
    >
      <Form
        className="CollectFromStore-search"
        noValidate={true}
        onSubmit={[Function]}
      >
        <Connect(UserLocatorInput)
          selectedCountry="United Kingdom"
          submitHandler={[Function]}
        />
      </Form>
    </div>
    <div
      className="CollectFromStore-columnRight"
    />
  </div>
  <div
    className="CollectFromStore-espotContainer"
  />
</div>
`;

exports[`<CollectFromStore/> @renders when stores is empty 1`] = `
<div
  className="CollectFromStore"
>
  <h3
    className="CollectFromStore-title"
  >
    Where do you want to collect from?
  </h3>
  <div
    className="CollectFromStore-row"
  >
    <div
      className="CollectFromStore-columnLeft"
    >
      <Form
        className="CollectFromStore-search"
        noValidate={true}
        onSubmit={[Function]}
      >
        <Connect(UserLocatorInput)
          selectedCountry="United Kingdom"
          submitHandler={[Function]}
        />
      </Form>
    </div>
    <div
      className="CollectFromStore-columnRight"
    />
  </div>
  <div
    className="CollectFromStore-espotContainer"
  />
</div>
`;
